internal/cache/cache.go

```diff
@@ -0,0 +1,199 @@
+	APIToken string
+}
+
+// CacheClient defines the interface for cache operations
+type CacheClient interface {
+	Save(ctx context.Context, cacheID string) (zstash.SaveResult, error)
+	Restore(ctx context.Context, cacheID string) (zstash.RestoreResult, error)
+	ListCaches() []cache.Cache
+}
+
+// Save saves caches based on the provided configuration and logs results as each cache is processed
+func Save(ctx context.Context, l logger.Logger, cfg Config) error {
+	cacheClient, cacheIDs, err := setupCacheClient(ctx, l, cfg)
+	if err != nil {
+		return err
+	}
+
+	if cacheClient == nil {
+		l.Info("No caches defined in the cache configuration file, nothing to save")
+		return nil
+	}
+
+	return saveWithClient(ctx, l, cacheClient, cacheIDs)
+}
+
+// Restore restores caches based on the provided configuration and logs results as each cache is processed
+func Restore(ctx context.Context, l logger.Logger, cfg Config) error {
+	cacheClient, cacheIDs, err := setupCacheClient(ctx, l, cfg)
+	if err != nil {
+		return err
+	}
+
+	if cacheClient == nil {
+		l.Info("No caches defined in the cache configuration file, nothing to restore")
+		return nil
+	}
+
+	return restoreWithClient(ctx, l, cacheClient, cacheIDs)
+}
+
+// loadCacheConfiguration loads cache configuration from a YAML file
+func loadCacheConfiguration(cacheConfigFile string) ([]cache.Cache, error) {
+	data, err := os.ReadFile(cacheConfigFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read cache config file: %w", err)
+	}
+
+	var caches []cache.Cache
+	if err := yaml.Unmarshal(data, &caches); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal cache config file: %w", err)
+	}
+
+	return caches, nil
+}
+
+// setupCacheClient creates a cache client and determines which cache IDs to process
+func setupCacheClient(ctx context.Context, l logger.Logger, cfg Config) (*zstash.Cache, []string, error) {
+	client := api.NewClient(ctx, version.Version(), cfg.APIEndpoint, cfg.APIToken)
+
+	caches, err := loadCacheConfiguration(cfg.CacheConfigFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load cache configuration: %w", err)
+	}
+
+	if len(caches) == 0 {
+		return nil, nil, nil
+	}
+
+	cacheClient, err := zstash.NewCache(zstash.Config{
+		Client:       client,
+		BucketURL:    cfg.BucketURL,
+		Format:       "zip",
+		Branch:       cfg.Branch,
+		Pipeline:     cfg.Pipeline,
+		Organization: cfg.Organization,
+		Caches:       caches,
+		OnProgress: func(stage, message string, current, total int) {
+			l.WithFields(
+				logger.StringField("stage", stage),
+				logger.StringField("message", message),
+				logger.IntField("current", current),
+				logger.IntField("total", total),
+			).Info("Cache progress")
+		},
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create cache client: %w", err)
+	}
+
+	// Determine which cache IDs to process
+	var cacheIDs []string
+	if cfg.Ids != "" {
+		cacheIDs = strings.Split(cfg.Ids, ",")
```

Comment:
Do we need to assert that the cache exist for a cache id?

Reply:
Good point fixed.